Add TLS CLI commands (#41418)
quarkusio/quarkus: docs/src/main/asciidoc/tls-registry-reference.adoc

%prod.quarkus.http.insecure-requests=disabled
----

== Quarkus CLI commands and development CA (Certificate Authority)

The TLS registry provides CLI commands to generate a development CA and trusted certificates.
This avoids having to use self-signed certificates locally.

[source, shell]
----

> > quarkus tls-registry

Reviewer: The config namespace is `quarkus.tls`. Do we want to be consistent at the CLI level and have `quarkus tls`? It might end up being a source of conflict though so I would understand if you thought it wouldn't be a good idea.

Author: Unfortunately, it is not possible to configure the "tls-registry" part when implementing a CLI plugin. It's the extension name that is reused (and it does not seem to be configurable). We can rename the extension, but that would be a massive change.

Author: Well, if you know a way, I'm very interested! 

@iocanel maybe?

Reviewer 2: I recently pushed a change that is supposed to allow for aliasing. I will comment below with how its meant to be done.

Author: I set the alias to `tls`, so much better!

Install and Manage TLS development certificates
Usage: tls-registry [COMMAND]

> Commands:

Author: Good question. What's the recommendation?

> Commands:

Reviewer: I think I would keep it as it is for now. They are fundamentally different operations.

  generate-quarkus-ca   Generate Quarkus Dev CA certificate and private key.
  generate-certificate  Generate a TLS certificate with the Quarkus Dev CA if
                          available.
----

> In most case, you generate the Quarkus Development CA once, and then generate certificates signed by this CA.

Reviewer: ```suggestion
In most cases, you generate the Quarkus Development CA once, and then generate certificates signed by this CA.
```

The Quarkus Development CA is a Certificate Authority that can be used to sign certificates locally.
It is only valid for development purposes and only trusted on the local machine.
The generated CA is located in `$HOME/.quarkus/quarkus-dev-root-ca.pem`, and installed in the system trust store.

=== CA, signed vs. self-signed certificates

When developing with TLS, you can use two types of certificates:

> - a self-signed certificate: the certificate is signed by the same entity that uses it. It is not trusted by default. It's generally what we uses when we don't have a CA, or don't want to dig too much into TLS. This is not a production setup, just for development.

Reviewer: ```suggestion
- a self-signed certificate: the certificate is signed by the same entity that uses it. It is not trusted by default. It's generally what we use when we don't have a CA, or don't want to dig too much into TLS. This is not a production setup, and may be used only for development.
```

- a signed certificate: the certificate is signed by a Certificate Authority (CA). The CA is a trusted entity that signs the certificate. The certificate is trusted by default. This is what we use in production.

We could use self-signed certificate when running application locally, but it's not always convenient.
Typically, browsers will not trust the certificate, and you will have to import it manually.
`curl`, `wget`, `httpie` and other tools will also not trust the certificate.

To avoid this, we can use a development CA to sign the certificates, and install it into the system trust store.
Thus, every certificate signed by this CA will be trusted by the system.

Quarkus makes it easy to generate a development CA and certificates signed by this CA.

=== Generate a development CA

The development CA is a Certificate Authority that can be used to sign certificates locally.
Note that the generated CA is only valid for development purposes, and only trusted on the local machine.

To generate a development CA, use the following command:

[source, shell]
----

> quarkus tls-registry generate-ca-certificate --install --renew --trusstore

Reviewer: ```suggestion
quarkus tls-registry generate-ca-certificate --install --renew --truststore
```

The generated CA is located in `$HOME/.quarkus/quarkus-dev-root-ca.pem`.

WARNING: When installing the certificate, your system may ask for your password to install the certificate in the system trust store, or ask for confirmation in a dialog (on Windows).

IMPORTANT: On Windows, makes sure you run from an elevated terminal (run as administrator) to install the CA in the system trust store.

`--renew` renews the CA if it already exists.
When this option is used, you need to re-generate the certificates that were signed by the CA, as the private key is changed.
Note that if the CA expires, it will automatically be renewed (without passing `--renew`).

`--truststore` also generates a PKCS12 trust store containing the CA certificate.

=== Generate a trusted (signed) certificate

Once you have installed the Quarkus Development CA, you can generate a trusted certificate.
It will be signed by the Quarkus Development CA, and so trusted by your system.

[source, shell]
----
quarkus tls-registry generate-certificate --name my-cert
----

This generates a certificate signed by the Quarkus Development CA, and so if properly installed / imported, will be trusted by your system.

> The certificate is stored in `./.certs/`.

Author: The .gitignore only contains `.quarkus/cli/plugins/` It would not be the right place for the certs. So, we will have to modify the generated `.gitignore`.

> The certificate is stored in `./.certs/`.

Author: Added in the .gitignore.

> The certificate is stored in `./.certs/`.

Reviewer: Ah yes, sorry, I checked a project that I manually adjusted.

But I would probably go with `.quarkus/certs` and add it to the `.gitignore`.

Two files are generated:

- `$NAME-keystore.p12` - contains the private key and the certificate. It's password protected.
- `$NAME-truststore.p12` - contains the CA certificate, that you can used as trust store (for test, for instance).

More options are available:

[source, shell]
----
Usage: tls-registry generate-certificate [-hrV] [-c=<cn>] [-d=<directory>]
       -n=<name> [-p=<password>]
Generate a TLS certificate with the Quarkus Dev CA if available.
  -c, --cn=<cn>       The common name of the certificate. Default is 'localhost'
  -d, --directory=<directory>
                      The directory in which the certificates will be created.
                        Default is `.certs`
  -n, --name=<name>   Name of the certificate. It will be used as file name and
                        alias in the keystore
  -p, --password=<password>
                      The password of the keystore. Default is 'password'

>   -r, --renew         Whether existing certificates will need to be replaced

Author: you mean a CA you already own? I would need the private key. I would rather not do that for security reasons.

----

> When generating the certificate, a `.env` file is also generated making the Quarkus Dev mode aware of these certificates.

Reviewer: ```suggestion
When generating the certificate, a `.env` file is also generated making the Quarkus dev mode aware of these certificates.
```

So, then, if you run your application in dev mode, it will use these certificates:

[source, shell]
----

> mvn quarkus:dev

Reviewer: ```suggestion
./mvnw quarkus:dev
```

...
INFO  [io.quarkus] (Quarkus Main Thread) demo 1.0.0-SNAPSHOT on JVM (powered by Quarkus 999-SNAPSHOT) started in 1.286s. Listening on: http://localhost:8080 and https://localhost:8443
----

> Now, you can open the dev ui using HTTPS: `https://localhost:8443/q/dev`, or issue a request using `curl`:

Reviewer: ```suggestion
Now, you can open the Dev UI using HTTPS: `https://localhost:8443/q/dev`, or issue a request using `curl`:
```
